server/envconfig/config.go

```diff
@@ -163,12 +182,19 @@ func LoadConfig() {
 		}
 	}
 
-	if onp := os.Getenv("OLLAMA_MAX_QUEUE"); onp != "" {
+	if onp := clean("OLLAMA_MAX_QUEUE"); onp != "" {
 		p, err := strconv.Atoi(onp)
 		if err != nil || p <= 0 {
 			slog.Error("invalid setting", "OLLAMA_MAX_QUEUE", onp, "error", err)
 		} else {
 			MaxQueuedRequests = p
 		}
 	}
+
+	CudaVisibleDevices = clean("CUDA_VISIBLE_DEVICES")
```

Comment:
@dhiltgen is this public variable `CudaVisibleDevices` meant to be used in `cudaGetVisibleDevicesEnv`?
https://github.com/ollama/ollama/blob/ee49844d09c5ee76a258ac53971e37bbf7857400/gpu/cuda_common.go#L10

Reply:
Good question.  

For NVIDIA GPUs, we use their C libraries to discover the GPUs.  We've recently switched to using the Driver API as our primary source, and fall back to the cuda runtime library if that fails (most likely we'll remove the cuda runtime code in a few releases as long as the Driver API looks reliable)  Those libraries already implement filtering based on this environment variable, so the calls [here ](https://github.com/ollama/ollama/blob/main/gpu/gpu_info_nvcuda.c#L87) and [here ](https://github.com/ollama/ollama/blob/main/gpu/gpu_info_cudart.c#L85) will only return a subset of GPUs based on that variable if the user has set it.

In our scheduler, we pick which GPU (of the GPUs exposed) to run a model on, and when we run the subprocess for inference, we wire up the environment variable to ensure that subprocess uses exactly the set of GPUs we want it to use.